### ISSUE TITLE
Improve setup process

### ISSUE THREAD

**Issue by Klaudioz** [karlicoss/promnesia#85]
I just discovered this project depends in a way of another projects created by you: HPI and kython.

So, I'm doing a test and I replaced the config.py.example to the config found on: https://github.com/karlicoss/promnesia/blob/master/doc/config.py

So I got this:
```
klaudioz@wsb-server:~/promnesia$ ./scripts/promnesia demo --port 13131 --config config.py
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/runpy.py", line 192, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.8/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/home/klaudioz/promnesia/src/promnesia/__main__.py", line 180, in <module>
    main()
  File "/home/klaudioz/promnesia/src/promnesia/__main__.py", line 173, in main
    do_demo(index_as=getattr(args, 'as'), params=args.params, port=args.port, config_file=args.config)
  File "/home/klaudioz/promnesia/src/promnesia/__main__.py", line 100, in do_demo
    config.load_from(config_file)
  File "/home/klaudioz/promnesia/src/promnesia/config.py", line 56, in load_from
    instance = import_config(config_file)
  File "/home/klaudioz/promnesia/src/promnesia/config.py", line 72, in import_config
    spec.loader.exec_module(mod) # type: ignore
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "config.py", line 26, in <module>
    from promnesia.sources import takeout, instapaper, pocket, fbmessenger, twitter, roamresearch, hypothesis
  File "/home/klaudioz/promnesia/src/promnesia/sources/takeout.py", line 10, in <module>
    from my.kython.kompress import kexists, kopen
ModuleNotFoundError: No module named 'my.kython'; 'my' is not a package
```

After that I realized the existence of kython. I cloned the repository, I run: `pip install -r requirements.txt` but I don't know what else to do to make work promnesia with this config file.

Obviously I'm interested to use promnesia but with its maximum power, integrating it with the most of platforms I use.

**Comment by karlicoss** [comment]
Ah! Yep, I need to improve documentation for this.
These are optional dependencies for some sources, you can see them [here](https://github.com/karlicoss/promnesia/blob/52a719a2bbbb4492efa887d4770c9119ed15f39e/setup.py#L67-L83)

E.g. if you want to use `markdown` source, you need to run `pip3 install --user '.[markdown]'`.

In this case, takeout (and all packages that try to import `my.something`) is provided by [HPI](https://github.com/karlicoss/HPI#readme), so let me know if you can follow its setup guide.

**Comment by Klaudioz** [comment]
Yeah, probably it's a little more complex than expected.

After run this: `sudo pip3.8 install --user HPI dataset mistletoe orgparse beautifulsoup4 lxml pyjq`

I got the same message. Any chance to generate a Dockerfile to have the entire environment working ?. I can help with that, but I need to understand first how everything is working.

**Comment by karlicoss** [comment]
Odd.. okay, some things to troubleshoot in the meantime:

1. This should print `usage: Human Programming Interface...` and exit:  `python3.8 -m my --help`
2. Can you share the output of this? `python3.8 -m pip list | grep HPI` -- it should list the package and the version.


P.S. it's probably not a good idea to use `pip install --user` with sudo: you would end up with files owned by root in your user python directory (usually in `~/.local/lib/python3.8`), and also kind of unsafe to install random packages as root. You can find more info [here](https://stackoverflow.com/a/37104948/706389) and in related questions. Although I don't see it would make any difference to this particular issue.

-----
Regarding docker: yep, there is an [issue](https://github.com/karlicoss/promnesia/issues/55) for Docker setup.
I've been postponing this because running this with docker is a bit more complicated (you'd have to figure out cron, manage mounts, etc), but now I can see it can be valuable even without the data, merely for documentation purposes.

In terms of how everything is working -- I've been improving the docs lately, at the moment got a (somewhat overwhelming map) [here](https://beepb00p.xyz/myinfra.html#exports). Doesn't help in terms of setting up the environment, but in case you'd like a high level view.

For promnesia specifically I also had this diagram, need to make it a bit clearer though:

```
: twitter    data -> HPI  --> source.twitter    --\
: instapaper data -> HPI  --> source.instapaper ---\
: ....
: org-mode files          --> source.org        ----\
: markdown files          --> source.markdown   -----> promnesia database (server/backend)  ...  promnesia browser addon (frontend)
: HTML     files          --> source.html       ----/
: JSON     files          --> source.json       ---/
: plaintext (other) files --> source.plaintext  --/
```

**Comment by Klaudioz** [comment]
OMG, that diagram is a monstruosity .. 😂

About the commands: (very weird)
```
klaudioz@wsb-server:~$ python3.8 -m my --help
klaudioz@wsb-server:~$ python3.8 -m pip list | grep HPI
HPI               0.2.20200514
```

I uninstalled and purge python 3.5 but it's still showing that output.

**Comment by karlicoss** [comment]
All right, I reproduced exactly the same trace as you had on a blank virtual machine:
```
Traceback (most recent call last):
  File "/opt/hostedtoolcache/Python/3.8.2/x64/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
....
  File "config.py", line 26, in <module>
    from promnesia.sources import takeout, instapaper, pocket, fbmessenger, twitter, roamresearch, hypothesis
  File "/home/runner/work/pymplate/pymplate/promnesia/src/promnesia/sources/takeout.py", line 10, in <module>
    from my.kython.kompress import kexists, kopen
ModuleNotFoundError: No module named 'my'
```

After that I ran `pip3 install --user HPI` and reran:
```
runner@fv-az68:~/work/pymplate/pymplate/promnesia$ scripts/promnesia demo --port 13131 --config config.py
......
  File "/home/runner/.local/lib/python3.8/site-packages/my/instapaper.py", line 7, in <module>
    from my.config import instapaper as config
ImportError: cannot import name 'instapaper' from 'my.config' (/home/runner/.local/lib/python3.8/site-packages/my/config/__init__.py)
```

Which is what expected, because it wants a [private config](https://github.com/karlicoss/HPI/blob/master/doc/SETUP.org#private-configuration-myconfig) first.

Next, I uninstalled: `pip3 uninstall HPI` and installed again with `sudo pip3 install --user HPI` as you did, and it also didn't work. 

**TLDR**: the best is to uninstall first: `sudo pip3 uninstall HPI` and then install again *without sudo*: `pip3 install --user HPI`. 

Longer explanation:

```
# without sudo -- PIP uses 'correct' version
runner@fv-az68:~/work/pymplate/pymplate/promnesia$ which pip3
/opt/hostedtoolcache/Python/3.8.2/x64/bin/pip3

# with sudo -- falls back to the system version (which is python3.5 in your case, and python3.6 for me)
runner@fv-az68:~/work/pymplate/pymplate/promnesia$ sudo which pip3
/usr/bin/pip3
```

And even weirder:

```
runner@fv-az68:~/work/pymplate/pymplate/promnesia$ python3.6 -m my --help
/usr/bin/python3.6: No module named my
runner@fv-az68:~/work/pymplate/pymplate/promnesia$ sudo python3.6 -m my --help
usage: Human Programming Interface [-h] {hello} ...
...
```

Even though you purged, it's possible it's falling back to python2, or something. Not sure it's even safe to purge default python version (unless you know what are you doing) because many operating system scripts are written in Python.

I have to admit, I don't understand exactly the way packages are resolved when some of them are installed with `sudo pip`, but it's best not to use sudo with pip anyway.

**Comment by Klaudioz** [comment]
Thanks for your detailed explanation. Even when I tried to fixed with what you put in ypur last message didn't work, so I decided start from zero creating a new machine with Ubuntu 20.03 in he cloud. So what I did:

```
sudo apt-get update
sudo apt-get install python3-pip
git clone https://github.com/karlicoss/promnesia.git && cd promnesia
# Rename config file and replace the content with https://raw.githubusercontent.com/karlicoss/promnesia/master/doc/config.py
pip3 install --user .
pip3 install --user logzero dataset HPI
cd ..
git clone https://github.com/karlicoss/HPI.git
mkdir -p ~/.config/my
cp -R ~/repos/HPI/my ~/.config/my
```

What I'm getting:
```
python3 -m my --help
usage: Human Programming Interface [-h] {hello} ...

positional arguments:
  {hello}
    hello     TODO just a stub, remove later

optional arguments:
  -h, --help  show this help message and exit

Tool for HPI. Work in progress, will be used for config management, troubleshooting & introspection
```

```
admin03@promnesia:~/repos/promnesia$ ./scripts/promnesia demo --port 13131 --config config.py
Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 193, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/admin03/repos/promnesia/src/promnesia/__main__.py", line 180, in <module>
    main()
  File "/home/admin03/repos/promnesia/src/promnesia/__main__.py", line 173, in main
    do_demo(index_as=getattr(args, 'as'), params=args.params, port=args.port, config_file=args.config)
  File "/home/admin03/repos/promnesia/src/promnesia/__main__.py", line 100, in do_demo
    config.load_from(config_file)
  File "/home/admin03/repos/promnesia/src/promnesia/config.py", line 56, in load_from
    instance = import_config(config_file)
  File "/home/admin03/repos/promnesia/src/promnesia/config.py", line 72, in import_config
    spec.loader.exec_module(mod) # type: ignore
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "config.py", line 26, in <module>
    from promnesia.sources import takeout, instapaper, pocket, fbmessenger, twitter, roamresearch, hypothesis
  File "/home/admin03/repos/promnesia/src/promnesia/sources/instapaper.py", line 5, in <module>
    import my.instapaper as ip
  File "/home/admin03/repos/HPI/my/instapaper.py", line 7, in <module>
    from my.config import instapaper as config
ImportError: cannot import name 'instapaper' from 'my.config' (/home/admin03/.config/my/my/config/__init__.py)
```

Previously I had the same last one but plus this: 
```
/home/admin03/repos/HPI/my/core/init.py:44: UserWarning: my.config package isn't found! (expected at /home/admin03/.config/my). This is likely to result in issues.
  warnings.warn(f"my.config package isn't found! (expected at {mycfg_dir}). This is likely to result in issues.")
```
But I fixed it with the last two lines from my script.
So, my understanding of this is promnesia needs HPI, and HPI has an own package system, so probably I need to add the HPI packages like instapaper, telegram, etc.

**Comment by karlicoss** [comment]
The last log actually looks good, means we're on the right path! 

HPI doesn't have a packages system (not a good idea to reinvent from scratch), I am thinking about something slim, but on top of existing mechanisms like PIP, but so far have been avoiding it. The only 'packaging' it's relying on are python's builtin mechanisms:

- "namespace packages", which are simply based on the filesystem layout
- dynamic imports, i.e. taking arbitrary path as a string an importing

The bit that's failing at the moment is actually missing config (which is in particular implemented as a package too, I'm [expanding on that here](https://github.com/karlicoss/HPI/blob/master/doc/CONFIGURING.org)). 

I guess error messages could be improved, e.g. I could link to this [section of the setup guide](https://github.com/karlicoss/HPI/blob/master/doc/SETUP.org#private-configuration-myconfig) from the warning. Let me know if you can follow it!

It's probably easiest if you start with one data source, and setting up the rest should be similar. I've just updated the docs on modules with few more examples: [here](https://github.com/karlicoss/HPI/blob/master/doc/MODULES.org) -- let me know if anything there is confusing. I'd say the easiest to start with is probably google takeout as it only requires you to have takeout archives.

**Comment by Klaudioz** [comment]
Thank you. I made it work just using the basic config.py file but I don't know how powerful can be with that file. I can see the output showing things but the promnesia icon is always empty. Even going to: chrome-extension://blablabla/search.html is empty. That URL I'd to figure out it because no shortcut is working in my two laptops(Chrome with Linux and Mac) so I went to options and I replaced options_page.html with search.html
I've used Telegram, Twitter, Roam and Hackernews but I can't see any feature of promnesia yet.

**Comment by karlicoss** [comment]
Empty as in the page is completely empty, or you get "Found 0 visits" when you search? 

If it's the former can you share which version of chrome you're using? I'm testing against chrome, so expect it to work...

Couple more things to troubleshoot, on the options page:
- is "check backend status" working? It should respond with "OK" and the path to your database.
- another thing you could try is to tick "enable verbose errors" -- and let me know if anything interesting pops up when you try using promnesia

**Comment by Klaudioz** [comment]
Check backend status is working and I have selected the "enable verbose errors"
I'm using Chrome 81.0.4044.138 (Official Build) (64-bit) in Ubuntu 18.04

If I see my logs I can see normally:
```
[DEBUG   2020-05-18 15:03:34,620 promnesia server.py:137] got 0 visits from db
[DEBUG   2020-05-18 15:03:34,620 promnesia server.py:148] responding with 0 visits
```

I can go to HackerNews, go to one link, close it and repeat again and always I got 0 in the log. If I turn down the server I got errors from Chrome extension immediately.

I've attached my logs.
[nohup.zip](https://github.com/karlicoss/promnesia/files/4645045/nohup.zip)

**Comment by karlicoss** [comment]
Ah, gotcha. I guess couple of things:
- if the database was empty in the first place, it wouldn't get any visits, it's read only from the extension's perspective. Perhaps I should clarify it in the readme!
- the 'local' (i.e. your chrome history) is always retrieved from the browser itself, but there was an artifact left from ancient times, which [added a 30 minutes delay](https://github.com/karlicoss/promnesia/blob/88098d5b1fb8ef70ba2fda6d3b5152833afb33bb/extension/src/background.js#L140) for actually showing them. It's not in the new version of the extension, but the chrome version is still being reviewed, hopefully won't take too long.

So, to actually make promnesia useful you'd need to populate the database first using some of the sources. 

Your log looks good, the only error is `/usr/share/doc/python3/html/faq` -- it's just a placeholder and apparently isn't present on you system; so perhaps you can try running it against some directory with actually useful data (org-mode/markdown/html). 
You can try exactly the same command as [here](https://github.com/karlicoss/promnesia#try-it-out), just to double check that the extension is working for you.

Or you could set up some other data source (e.g. from HPI). Let me know if you have anything specific in mind.

**Comment by karlicoss** [comment]
By the way, I updates some docs on modules & dependencies, hopefully makes a bit more sense now: [docs/SOURCES](https://github.com/karlicoss/promnesia/blob/master/doc/SOURCES.org)

**Comment by Klaudioz** [comment]
I'll try it again from zero this weekend. Thanks for you help

**Comment by karlicoss** [comment]
btw, added more docs for HPI, which will hopefully make Promnesia setup clearer as well!: ["Data flow"](https://github.com/karlicoss/HPI/blob/master/doc/SETUP.org#data-flow)
